_posts/2018-09-26-artsy-frontend-history.markdown

```diff
@@ -0,0 +1,176 @@
+views and models dealt with backend communication and notifying the view of state changes. For CSS, we used the
+SASS CSS preprocessor. The Rails backend served initial pages that were then populated with follow-up API calls
+made on the client-side. At a _very_ high level, this isn't _that_ different from what we do today with React.
+
+Our site was built with a framework called [Backbone][], which was really well-suited for our needs at the time.
+From their documentation:
+
+> Philosophically, Backbone is an attempt to discover the minimal set of data-structuring (models and collections)
+> and user interface (views and URLs) primitives that are generally useful when building web applications with
+> JavaScript. In an ecosystem where overarching, decides-everything-for-you frameworks are commonplace, and many
+> libraries require your site to be reorganized to suit their look, feel, and default behavior — Backbone should
+> continue to be a tool that gives you the _freedom_ to design the full experience of your web application.
+
+As an outsider to the web at that time, I can't comment too heavily on Backbone. It seems like the freedom
+(emphasis theirs) that they describe is a freedom from tangled jQuery code everywhere. I think our definition of
+freedom on the web frontend has evolved since then, but that's just my feeling.
+
+The other key component to our web frontend was [CoffeeScript][]. According to its documentation, "CoffeeScript is
+a little language that compiles into JavaScript", which was pretty important at the time. JavaScript in 2011 was
+very different from JavaScript today. The CoffeeScript docs also state that "JavaScript has always had a gorgeous
+heart", which I'm not sure I'd agree with to be honest, but the CoffeeScript project really shows how a handful of
+engineers working to improve something they care about can change an entire industry. While I don't think
+contemporary JavaScript would have gotten as good as it has without CoffeeScript, it's a bit anachronistic to see
+it used today.
+
+Our goal as a (very small!) engineering team at the time was to keep our moving parts to a minimum.
+Rails+SASS+CoffeeScript+Backbone helped us achieve that goal, and we couldn't have gotten this far without the help
+of those projects.
+
+## Ezel & Friends
+
+In November 2013, we split our web frontend from the API backend. You can read
+[all the details in this blog post](2013_review), but the story is summarized nicely as "moving from a single
+monolithic application to modular Backbone apps that run in Node and the browser and consume our external API."
+This move from monolith to modular systems continues to influence day-to-day work on the Artsy Engineering team.
+
+We had already started moving away from a typical Rails app by moving our API to [Grape][] in order to support an
+iOS application. The monolith also had some clear drawbacks including SEO problems, severe page load times,
```

Comment:
Thanks Joey – I'll revisit adding trade-offs to this post, since I wanted this to discuss them and have a kind of "this worked for us but might not for everyone" feel. 

For this specific feedback, I drew from [Craig's blog post on Ezel](http://artsy.github.io/blog/2013/11/30/rendering-on-the-server-and-client-in-node-dot-js/), where he said:

> We knew we were outgrowing this monolithic project because we had some clear problems...
> 
> - ...
> - SEO issues like building escaped fragment pages in Ruby on the server while our users saw what JavaScript rendered on the client.

Do you think this paragraph needs _more_ detail, or less?

Reply:
🤔I probably read that at the time but I honestly don't know what that bullet is talking about. We had to pre-render pages both before and after the Force extraction, and if anything it was made much harder by putting routing and page logic in a separate system than the data.

Reply:
I wonder if this was all this stuff: http://artsy.github.io/blog/2012/06/25/replacing-hashbang-routes-with-pushstate/

Reply:
:ceiling-cat: I concur with @joeyAghion’s assessment that usually these are advantages of a monolith.

Reply:
Cool, that makes sense. Maybe I'll take out the SEO mention, but I'd like to link to this GitHub thread from somewhere in the post (to show how context around code gets lost to time, and why documenting our decisions is important).